Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -259,27 +266,35 @@ extension AnyMessageStorage {
       return false
     }
 
-    // If we have both data's, compare those.
-    // The best option is to decode and compare the messages; this
-    // insulates us from variations in serialization details.  For
-    // example, one Any might hold protobuf binary bytes from one
-    // language implementation and the other from another language
-    // implementation.  But of course this only works if we
-    // actually know the message type.
-    //if let myMessage = _message {
-    //    if let otherMessage = other._message {
-    //        ... compare them directly
-    //    } else {
-    //        ... try to decode other and compare
-    //    }
-    //} else if let otherMessage = other._message {
-    //    ... try to decode ourselves and compare
-    //} else {
-    //    ... try to decode both and compare
-    //}
-    // If we don't know the message type, we have few options:
-    // If we were both deserialized from proto, compare the binary value:
-    // If we were both deserialized from JSON, compare content of the JSON?
+    // Since the library does lazy Any decode, equality is a very hard problem.
+    // It things exactly match, that's pretty easy, otherwise, one ends up having
+    // to error on saying they aren't equal.
+    //
+    // The best option would be to have Message forms and compare those, as that
+    // removes issues like map<> serialization order, some other protocol buffer
+    // implementation details/bugs around serialized form order, etc.; but that
+    // would also greatly slow down equality tests.
+    //
+    // Despite the problems with binary serialization, that may be the best bet
+    // as looking at most other language implementations around Any they don't
+    // do lazy decode and instead keep all the `value` fields in binary form and
+    // end up just comparing those binary fields when comparing Anys.
+
+    // If both have messages, check if they are the same.
```

Comment:
This seems to work.   It adds the heterogenous equality as a separate method on `Message`.  This avoids adding the type check to `==` but at the cost of adding another method to the `Message` protocol:
```
diff --git a/Sources/SwiftProtobuf/Message.swift b/Sources/SwiftProtobuf/Message.swift
index 518820a..b58df2e 100644
--- a/Sources/SwiftProtobuf/Message.swift
+++ b/Sources/SwiftProtobuf/Message.swift
@@ -112,8 +112,10 @@ public protocol Message: CustomDebugStringConvertible {
   /// debugging, for conformance with the `CustomDebugStringConvertible`
   /// protocol.
   var debugDescription: String { get }
+
+  func isEqualTo(message: Message) -> Bool
 }
 
 public extension Message {
 
   /// If the generated code needs to provide its own implementation, usually
@@ -197,6 +199,13 @@ public extension Message {
 
 public extension _MessageImplementationBase {
+  public func isEqualTo(message: Message) -> Bool {
+    if let other = message as? Self {
+      return self == other
+    }
+    return false
+  }
+
   public static func ==(lhs: Self, rhs: Self) -> Bool {
     return lhs._protobuf_generated_isEqualTo(other: rhs)
   }
 }

```